### PR TITLE
Add proposal for `%constructor%.construct()`

### PR DESCRIPTION
**Pull request by jasnell**
Proposal allows for old-style inheritance from new-style classes, e.g.

in Node.js:
```js
const util = require('util');

class A {
  constructor(a) {
    this.a = a;
  }
}

function B() {
  A.construct(this, 1);
}
util.inherits(B, A);

const b = new B();
console.log(b.a);      // 1
```

### CODE REVIEW

**Comment by loganfsmyth** [comment]
Maybe missing context on previous discussions, but wouldn't this already work via

```
function B() {
  return Reflect.construct(A, [ 1 ], new.target);
}
```

The proposal repo link is broken. It has a slash instead of a dash in it.

File: stage-0-proposals.md [hunk 1]
```diff
@@ -25,6 +25,7 @@ Stage 0 proposals are either
 |   | [Orthogonal classes](https://github.com/erights/Orthogonal-Classes)                                                                      | Mark S. Miller, Allen Wirfs-Brock | 0   |
 |   | [Normative ICU Reference](https://github.com/tc39/tc39-notes/blob/master/es8/2017-05/may-23.md#normative-icu-reference)                  | Domenic Denicola                | 0     |
 |   | [WHATWG URL](https://github.com/jasnell/proposal-url)                                                                                    | James M Snell                   | 0     |
+|   | [`%contstructor%.construct(thisArg[, arguments])`](https://github.com/jasnell/proposal/construct)                                        | James M Snell                   | 0     |
```
Comment (loganfsmyth) [inline stage-0-proposals.md]: Looks like a typo FYI, `proposal-construct`, not `proposal/construct`.

**Comment by jasnell** [comment]
No, `Reflect.construct(A, [a], new.target)` does not actually work in this case. Consider:

```js
const util = require('util');

class A {
  constructor(a) {
    this.a = a;
  }
}

function B(a) {
  Reflect.construct(A, [a], new.target);
}
util.inherits(B, A);

const b = new B(1);
console.log(b.a);

// Prints: undefined
```

**Comment by jasnell** [comment]
@loganfsmyth ... to be honest, I'm not 100% why I hadn't picked up on that slight variation previously. I think. I'm going to close this for now.

**Comment by jasnell** [comment]
Ah.. @loganfsmyth ... actually... here's the case where this breaks down (sorry, had to run an example to refresh my memory): 

```js
const util = require('util');

class A {
  constructor(a) {
    this.a = a;
  }
}

///

function B(a) {
  if (!(this instanceof B))
    return new B(a);
  return Reflect.construct(A, [a], new.target);
}
util.inherits(B, A);

B.prototype.foo = function() {};

// This all works fine
const b = B(1);
console.log(b.a);
console.log(b.foo);
console.log(b instanceof B);
console.log(b instanceof A);

///

// This, however, does not

function C() {
  if (!(this instanceof C))
    return new C();
  B.call(this, 2);
}
util.inherits(C, B);

const c = C();  // This dies with an error
console.log(c.a);
console.log(c.foo);
console.log(c instanceof C);
console.log(c instanceof B);
console.log(c instanceof A);
```

The reason this fails is because when `C()` calls `B.call(this)`, which is the old-style, `new.target` inside `B()` is `undefined`

**Comment by jasnell** [comment]
Btw, we hit this recently in Node.js core... which uses *many* old style classes that end users subclass for their own purposes. While we can have B subclass A successfully, C cannot subclass B without significant changes. The goal here is to find a mechanism that would allow us to update the implementation of these classes in Node.js core without breaking downstream code.

**Comment by allenwb** [comment]
`this instanceof X` is not the correct way to  determine whether a function has been "called as a constructor" rather than "called as a function" (ie, invoked via [[Constructor]] rather [[Call]])

The correct way is to test `new.target`.  So instead of:
```js
function B(...a) {
  if (!(this instanceof B))
    return new B(...a);
  return Reflect.construct(A, a, new.target);
}
```

you should do:
```js
function B(...a) {
  if (new.target===undefined) {
    //B was "called"
    return new B(...a);
  } 
  //"B was "constructed" 
  return Reflect.construct(A, a, new.target); 
}
```
Then to account for the possibility that B is being subclassed by an Node "old-style" class C elaborate B like::

```js
function B(...a) {
  if (new.target===undefined) {
    //B was "called"
    if (this==null)
       //B was normally called, treat it as an implicit: new B
       return new B(...a); 
    if (this instanceof B) 
      // B "super" constructed in the old manner, treat as implicit super construct 
      // with this as the new.target
      return Reflect.construct(A, a, this);
    //some other this hanky-panky
    throw Error("Invalid old-style subclass");
  } 
  //"B was "constructed" 
  return Reflect.construct(A, a, new.target); 
}
```

**Comment by bakkot** [comment]
@allenwb, `Reflect.construct(A, a, this);` doesn't do the right thing; the `new.target` parameter to `Reflect.construct` is meant to be the constructor function, not the instance being constructed. (And indeed if you try you get a [TypeError](https://tc39.github.io/ecma262/#sec-reflect.construct); see step 3.)

As far as I know, there's no way to invoke a superclass constructor on an arbitrary object, which is what `B` needs to do: that's how old-style "super" calls work.

**Comment by bakkot** [comment]
That said: `C` was always kinda broken, because `B` could have returned an arbitrary object instead of modifying `this`. If it were written instead as `return B.call(this, 2)`, you could make everything work:

```js
function B(...a) {
  if (new.target===undefined) {
    //B was "called"
    if (this==null)
       //B was normally called, treat it as an implicit: new B
       return new B(...a); 
    if (this instanceof B) {
      // B "super" constructed in the old manner, treat as implicit super construct 
      // with this as the new.target
      let rv = Reflect.construct(A, a, B);
      Object.setPrototypeOf(rv, Object.getPrototypeOf(this));
      return rv;
    }
    //some other this hanky-panky
    throw Error("Invalid old-style subclass");
  } 
  //"B was "constructed" 
  return Reflect.construct(A, a, new.target); 
}
```

**Comment by jasnell** [comment]
Yep, unfortunately there's a whole lot of code out there that relies on the somewhat broken old way that we simply cannot break. Nor can we require that users change. Until this is resolved, we simply cannot fully adopt classes in Node.js core for any existing API.

**Comment by allenwb** [comment]
Dealing with this is complicated by the fact that currently class constructors can not "called", but only constructed. So you are limited as to what you could do to within actual core class definition to remediate the problem.

But, doesn't your backwards compat issue only exist in situations where the node `util.inherits` is used?  What if `util.inherits(B,A)` recognizes  when A is a ES class constructor and in that cases inserts on the constructor side proto chain between B and A a synthetic constructor function similar to the one in https://github.com/tc39/proposals/pull/51#issuecomment-309148928

That synthetic constructor bridges between the two instantiations styles (assuming the old-style subclass reasonably conforms to the common conventions)

**Comment by jasnell** [comment]
That doesn't quite cover it because C would still be broken. In the old style, this is constructed at C, then passed in to B and essentially composed by walking up the supers. Even if we intercepted the class constructor in util.inherits we'd run into exactly the same problem. I'm going to stew on it a bit more to see if there's another possible solution here, but essentially, to solve the issue, we'd need a way of passing this into the constructor.

**Comment by loganfsmyth** [comment]
Passing `this` to a class constructor is a no-go isn't it? The constructor itself needs to be the thing creating the object in order to implement subclassing of constructors that create exotic objects with internal slots.

**Comment by leobalter** [comment]
We're establishing a weird pattern here. Proposals are being discussed in the PRs to add them as Stage 0, at this point, I just want to merge but I'm afraid to interrupt the discussion, but I believe it's good to add it to the list anyway for historical purposes if anything.

I'll try to work on an idea to bring until the next meeting to improve these discussions. In the meanwhile: @jasnell can we get this merged?

**Comment by jasnell** [comment]
Yes, absolutely :)

**Pull request merged**